spark_log_parser/loaders/__init__.py

```diff
@@ -0,0 +1,331 @@
+import abc
+import gzip
+import tarfile
+import logging
+
+from io import BufferedIOBase
+from pathlib import Path
+from typing import Iterator
+from urllib.parse import ParseResult
+
+from pydantic import BaseModel
+from aiodataloader import DataLoader
+from stream_unzip import stream_unzip
+
+FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
+
+logger = logging.getLogger("Loaders")
+
+
+class ArchiveExtractionThresholds(BaseModel):
+    entries = 100
+    size = 5000000000
+
+
+class ArchiveTooLargeError(AssertionError):
+    """Raised when we encounter an archive that expands to beyond the allowed filesize limit"""
+    pass
+
+
+class ArchiveTooManyEntriesError(AssertionError):
+    """Raised when we encounter an archive that has more entries in it than is allowable"""
+    pass
+
+
+class FileChunkStreamWrapper:
+    """
+    Minimal file-like wrapper to wrap the output of file streams that may be lacking methods like read/readlines (or in
+    other words, file streams that are raw generators). This allows us to wrap things like the output of `stream_unzip`
+    or botocore.StreamingBody.iter_lines() and pass those streams around as file-like objects.
+
+    Note - this wrapper does not implement the full file-like object API. Instead, it only implements at the moment
+        those methods that are actually called when loading files in various contexts. This doesn't mean that
+        new methods won't need to be implemented in the future as use-cases grow, it just means that at the moment
+        this minimal implementation satisfies the current use-cases we do have
+
+    Note #2 - once `chunks` is consumed, calls to read/lines will return empty byte buffers
+    """
+
+    total_size: int = 0
+
+    _maximum_allowed_size: ArchiveExtractionThresholds
+    _trailing_data: bytes = b''
+    _chunks: Iterator[bytes]
+
+    def __init__(self, chunks: Iterator[bytes], maximum_file_size: int = None):
+        assert chunks
+        self._chunks = chunks
+        self._maximum_allowed_size = maximum_file_size
+
+    def __iter__(self):
+        for chunk in self._chunks:
+            if not chunk:
+                continue
+            else:
+                self._add_chunk_to_size(chunk)
+                yield chunk
+
+    def _add_chunk_to_size(self, chunk):
+        self.total_size += len(chunk)
+        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+            raise ArchiveTooLargeError()
+
+    def read(self, size=-1) -> bytes:
+        content = self._trailing_data
+        self._trailing_data = b''
+
+        if not size or size == -1:
```

Comment:
This is an interesting point - I went with `-1` because that appears to be the default value in much of the standard library (e.g. - https://docs.python.org/3/library/io.html#io.RawIOBase.read). `None` would still probably work as a default, but we may want to keep the `-1` behaviour just to be consistent with the larger ecosystem. That said, also probably worth adding a check if `size < -1`, and raise an Error if it is